File: 1095.java
interface MountainArray {
    public int get(int index) {}

    public int length() {}
}

class Solution {
    public int findInMountainArray(int target, MountainArray mountainArr) {
        int n = mountainArr.length();
        int left = 0;
        int right = n - 1;
        int mid;
        int res = -1;
        while (left < right) {
            mid = left + (right - left) / 2;
            if (mountainArr.get(mid) > mountainArr.get(mid + 1)) {
                res = mid;
                right = mid - 1;
            }
            if (mountainArr.get(mid) < mountainArr.get(mid + 1))
                left = mid + 1;
        }

        int x = bsAscending(mountainArr, target, 0, res - 1);
        int y = bsDescending(mountainArr, target, res, n - 1);
        if (x == -1 || y == -1)
            return Math.max(x, y);
        else
            return Math.min(x, y);
    }

    public int bsAscending(
            MountainArray mountainArr, int target, int left, int right) {
        int mid;
        while (left <= right) {
            mid = left + (right - left) / 2;
            if (mountainArr.get(mid) == target)
                return mid;
            else if (target < mountainArr.get(mid))
                right = mid - 1;
            else
                left = mid + 1;
        }
        return -1;
    }

    public int bsDescending(MountainArray mountainArr, int target, int left, int right) {
        int mid;
        while (left <= right) {
            mid = left + (right - left) / 2;
            if (mountainArr.get(mid) == target)
                return mid;
            else if (target < mountainArr.get(mid))
                left = mid + 1;
            else
                right = mid - 1;
        }
        return -1;
    }
}

File: Python/1095.py
# """
# This is MountainArray's API interface.
# You should not implement it, or speculate about its implementation
# """
# class MountainArray:
#    def get(self, index: int) -> int:
#    def length(self) -> int:


class Solution:
    def findInMountainArray(self, target: int, mountainArr: "MountainArray") -> int:
        length = mountainArr.length()

        # Find Peak
        l, r = 1, length - 2
        while l <= r:
            m = l + (r - l) // 2
            left, mid, right = (
                mountainArr.get(m - 1),
                mountainArr.get(m),
                mountainArr.get(m + 1),
            )
            if left < mid < right:
                l = m + 1
            elif left > mid > right:
                r = m - 1
            else:
                break
        peak = m

        # Search left portion
        l, r = 0, peak - 1
        while l <= r:
            m = l + (r - l) // 2
            val = mountainArr.get(m)
            if val < target:
                l = m + 1
            elif val > target:
                r = m - 1
            else:
                return m

        # Search right portion
        l, r = peak, length - 1
        while l <= r:
            m = l + (r - l) // 2
            val = mountainArr.get(m)
            if val > target:
                l = m + 1
            elif val < target:
                r = m - 1
            else:
                return m

        return -1
